lib/redis/connection/ruby.rb

```diff
@@ -49,43 +49,47 @@ def gets
       end
 
       def _read_from_socket(nbytes)
-        begin
-          read_nonblock(nbytes)
-        rescue IO::WaitReadable
-          if IO.select([self], nil, nil, @timeout)
-            retry
-          else
-            raise Redis::TimeoutError
-          end
-        rescue IO::WaitWritable
-          if IO.select(nil, [self], nil, @timeout)
-            retry
-          else
-            raise Redis::TimeoutError
+        loop do
+          case chunk = read_nonblock(nbytes, exception: false)
+          when :wait_readable
+            unless IO.select([self], nil, nil, @timeout)
+              raise Redis::TimeoutError
+            end
+          when :wait_writable
+            unless IO.select(nil, [self], nil, @timeout)
+              raise Redis::TimeoutError
+            end
+          when nil
+            raise Errno::ECONNRESET
+          when String
+            return chunk
           end
         end
-      rescue EOFError
-        raise Errno::ECONNRESET
       end
 
       def _write_to_socket(data)
-        begin
-          write_nonblock(data)
-        rescue IO::WaitWritable
-          if IO.select(nil, [self], nil, @write_timeout)
-            retry
-          else
-            raise Redis::TimeoutError
-          end
-        rescue IO::WaitReadable
-          if IO.select([self], nil, nil, @write_timeout)
-            retry
-          else
-            raise Redis::TimeoutError
+        total_bytes_written = 0
+        loop do
+          case bytes_written = write_nonblock(data, exception: false)
+          when :wait_readable
+            unless IO.select([self], nil, nil, @timeout)
```

Comment:
Should this still use `@write_timeout` (and not `@timeout`)?

Reply:
Indeed. Good catch.

Reply:
Fixed in 087a11b585978cd3970d22d066b7d8ccd89a40f6